### DISCUSSION TITLE
Detecting a mastodon instance programmatically

### DISCUSSION THREAD

**Discussion by fuddl** [mastodon/mastodon#21493] [General]
I am working on a webextension that, among other things, collects mastodon addresses for wikidata. Since mastodon is decentralised with no definitive list of instances, I rely on regular expressions and existing addresses in order to detect if a url  represent something with a mastodon address.

For instance `https://example.com/@fuddl` looks like something that has the mastodon address `@fuddl@example.com`. Is there a test I could run against `example.com` that could tell me if it actually is?

My issue came up here:
https://github.com/fuddl/wd/issues/104

**Comment by trwnh** [comment]
1. Check `/.well-known/nodeinfo` for a links array, then get an `href` from one of the items
2. This will lead you to something like https://mastodon.social/nodeinfo/2.0 where you can check for `software.name`

**Comment by nightpool** [comment]
> detect if a url represent something with a mastodon address.

What do you mean by "Mastodon address" here? `https://example.com/@fuddl` is an ActivityPub actor URL, which is the canonical way to refer to users and is the unique identifier for a user's account. You can confirm this by making a GET request with an `Accept` header of `application/ld+json; profile="https://www.w3.org/ns/activitystreams"`, and seeing that the returned response has the "Person" (or "Group" or "Service") type. This is the recommended way to identify an ActivityPub actor. The @username@domain format is only a shorthand/convenience format used for user input / display, and is managed by a different spec (WebFinger)

The `.nodeinfo` method mentioned above is implemented by some servers, but is not related to ActivityPub and I would not recommend it.

**Comment by fuddl** [comment]
> What do you mean by "Mastodon address" here?

As described in [the wikidata property](https://www.wikidata.org/wiki/Property:P4033)

> address on the Mastodon decentralized social network. The form is: 'user@server.domain' there is no leading '@' as sometimes written to distinguish Mastodon addresses from email addresses.

To be clear, I only look for instances of mastodon, not any other ActivityPub node.

I was thinking of asking the api about which version the node runs on, If that would be a single request.

**Reply by trwnh** [reply]
There is no concept of a "Mastodon address" -- this address is based on Webfinger, and the Wikidata page seems nonsensical.

Why are you looking only for Mastodon? And do you mean Mastodon software or Mastodon API compatible?

**Reply by fuddl** [reply]
I think what they mean is something equivalent to a twitter handle, but on the mastodon network. I'm only trying to fill that field regardless of it makes sense 😅

**Reply by trwnh** [reply]
well there's no "Mastodon network" -- there is ActivityPub and the identifiers there are just HTTPS URLs that can be fetched with `Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"` as a header

the `@username@domain` comes from Webfinger and is just a pointer to the ActivityPub identifier. for example, getting `https://mastodon.social/.well-known/webfinger?resource=acct:trwnh@mastodon.social` will return a document pointing to `https://mastodon.social/users/trwnh`

**Comment by markcellus** [comment]
@fuddl, just checked out your linked issue https://github.com/fuddl/wd/issues/104. 

So in your example, if you want to check if the `about.me` in `khayrirrw@about.me` is a Mastodon address (aka ActivityPub URI), you'd just check if `https://about.me/.well-known/webfinger?resource=acct:khayrirrw@about.me` returns a 200 status with data. 

If there is no successful response, it's probably not a valid ActivityPub URI and likely not a Mastodon instance. But beware this may not work for clients other than Mastodon. But it sounds like you just only want to detect ActivityPub actors that are on Mastodon specifically, so this may work for you. You'd also have to deal with edge cases like whether to use www vs non-www subdomain, if the username might be different from the canonical username on the Mastodon instance and redirect isn't done properly, etc.

Hope this helps you a bit!

**Reply by fuddl** [reply]
this looks good to me. Thank you.

> You'd also have to deal with edge cases like whether to use www vs non-www subdomain, if the username might be different from the canonical username on the Mastodon instance and redirect isn't done properly, etc.

My assumption was, that if a mastodon instance is reachable under `www.example.com/@khayrirrw` then the handle would have to be `@khayrirrw@www.example.com` rather than `@khayrirrw@example.com`. Isn't that the case?
